### ISSUE TITLE
Best practice to handle large amount of overlays?

### ISSUE THREAD

**Issue by SMA-cgn** [openseadragon/openseadragon#2268]
Hi,

I would like to add > 1000 overlays to an image in order to just mark certain areas of the image. These overlays will not contain any content or so..

Is there any best practice or advise how to do this in regards of performance & maintainability?
Does it make sense to use built-in overlays?

Thanks a lot for your advice!

**Comment by iangilman** [comment]
What do you want these overlays to look like? Little dots, or rectangles, or other kinds of shapes? How would you feel about them getting bigger as you zoom in? 

For performance, I'm a fan of https://github.com/openseadragon/svg-overlay and https://github.com/joshua-gould/OpenSeadragonCanvasOverlayHd. 

The built-in overlays don't resize as you zoom in. SVG overlays do. Canvas overlays depend on you to do the drawing each frame, so it's entirely up to you.

Hard to say which would be best for your scenario... Maybe try all three and see which one provides the best performance?

**Comment by SMA-cgn** [comment]
Thanks a lot! For my purpose I just need humble div elements that carry an id. I will test the different options and provide feedback. :)

**Comment by iangilman** [comment]
You can add an event handler for the `zoom` event and watch for the zoom to go past your threshold. You can then add overlays at that time. Another option would be to already have the overlays added, but hidden (with `display: none` in their style or something) and you can change that based on the event.

**Comment by Aiosa** [comment]
For thousands of items, you would maybe want to consider `fabricJS` or `paperJS` overlays that are very performant in vector data rendering. https://openseadragon.github.io/#plugins

**Comment by SMA-cgn** [comment]
thanks a lot!

**Comment by pearcetm** [comment]
I agree with @Aiosa, both of these would be good choices. I'm not entirely sure of the existing plugin APIs, but the `fabricJS` and `paperJS` libraries both support importing directly from SVG. So, if you've already explored SVG as an option and have converted your annotations to that format already, it could (potentially) be fairly straightforward to give either/both of these a shot.

**Comment by SMA-cgn** [comment]
> You can add an event handler for the `zoom` event and watch for the zoom to go past your threshold. You can then add overlays at that time. Another option would be to already have the overlays added, but hidden (with `display: none` in their style or something) and you can change that based on the event. You could even just toggle a class on the page that triggers them all being displayed:
> 
> ```
> .my-overlay {
>   display: none;
> }
> 
> .zoomed-in-enough .my-overlay {
>   display: block;
> }
> ```

Thanks..  using the zooming event is very performant!
Is there also a way to add overlays depending on the viewport? Meaning if the coordinates of the overlay are in the current viewport.
If so, how would you do that?

Thanks a lot!

**Comment by Aiosa** [comment]
@SMA-cgn zooming event is just an event, is has no performance features, just a frequency. Checking for coordinate overlay being in current viewport can be done with 'update-viewport' event - a list of events can be viewed in the _documentation_ (https://openseadragon.github.io/docs/OpenSeadragon.Viewer.html). It seems to me like you want to add and remove overlays based on viewport position - in case you are trying to optimize your overlays this way, it is probably not the best idea. Aforementioned canvas libraries handle this in much clever ways. Just saying.

**Comment by iangilman** [comment]
To see if something is in the viewport, use `viewer.viewport.getBounds()` and compare against that. 

Depending upon what overlay system you're using, you might not need to bother with adding things only in the viewport... They may just be ignored anyway if they are outside the viewport. Of course, the only way to know for sure is to test it.
